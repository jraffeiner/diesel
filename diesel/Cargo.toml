[package]
name = "diesel"
version = "2.2.4"
license = "MIT OR Apache-2.0"
description = "A safe, extensible ORM and Query Builder for PostgreSQL, SQLite, and MySQL"
readme = "README.md"
documentation = "https://docs.rs/diesel/"
homepage = "https://diesel.rs"
repository = "https://github.com/diesel-rs/diesel"
keywords = ["orm", "database", "sql"]
categories = ["database"]
edition = "2021"
rust-version.workspace = true
include = [
    "src/**/*.rs",
    "tests/**/*.rs",
    "LICENSE-*",
    "README.md",
    "src/sqlite/connection/diesel_manage_updated_at.sql",
    "src/migration/setup_migration_table.sql",
]

[dependencies]
byteorder = { version = "1.0", optional = true }
chrono = { version = "0.4.20", optional = true, default-features = false, features = ["clock", "std", "serde"] }
chrono-tz = { version = "*", optional = true }
libc = { version = "0.2.0", optional = true }
mysqlclient-sys = { version = ">=0.2.5, <0.5.0",  optional = true }
mysqlclient-src = { version = "0.1.0", optional = true }
pq-sys = { version = ">=0.4.0, <0.8.0", optional = true }
pq-src = { version = "0.3", optional = true }
quickcheck = { version = "1.0.3", optional = true }
serde_json = { version = ">=0.8.0, <2.0", optional = true }
url = { version = "2.1.0", optional = true }
percent-encoding = { version = "2.1.0", optional = true }
uuid = { version = ">=0.7.0, <2.0.0", optional = true }
ipnetwork = { version = ">=0.12.2, <0.22.0", optional = true }
ipnet = { version = "2.5.0", optional = true }
num-bigint = { version = ">=0.2.0, <0.5.0", optional = true }
num-traits = { version = "0.2.0", optional = true }
num-integer = { version = "0.1.39", optional = true }
bigdecimal = { version = ">=0.0.13, < 0.5.0", optional = true }
bitflags = { version = "2.0.0", optional = true }
r2d2 = { version = ">= 0.8.2, < 0.9.0", optional = true }
itoa = { version = "1.0.0", optional = true }
time = { version = "0.3.9", optional = true, features = ["macros"] }
thiserror = { version = "*", optional = true }
bytes = { version = "*", optional = true }
enumflags2 = { version = "*", optional = true }
tracing = { version = "*", optional = true }
encoding_rs = { version = "*", optional = true }
connection-string = { version = "*", optional = true }
pretty-hex = { version = "*", optional = true }
native-tls = { version = "*", optional = true }
rustls = { version = "*", optional = true }
rustls-pemfile = { version = "*", optional = true }
rustls-native-certs =  { version = "*", optional = true }
rust_decimal =  { version = "*", optional = true }
winauth = { version = "*", optional = true }
rsfbclient = { version = "*", default-features = false, features = ["pure_rust"], optional = true }
downcast-rs = "1.2.1"

[dependencies.diesel_derives]
version = "~2.2.0"
path = "../diesel_derives"

[target.'cfg(not(all(target_family = "wasm", target_os = "unknown")))'.dependencies]
libsqlite3-sys = { version = ">=0.17.2, <0.32.0", optional = true, features = ["bundled_bindings"] }

[target.'cfg(all(target_family = "wasm", target_os = "unknown"))'.dependencies]
sqlite-wasm-rs = { version = ">=0.3.0, <0.4.0", optional = true, default-features = false }

[target.'cfg(all(target_family = "wasm", target_os = "unknown"))'.dev-dependencies]
# Something is dependent on it, we use feature to override it.
getrandom = { version = "0.2", features = ["js"] }
wasm-bindgen-test = "0.3.49"
sqlite-wasm-rs = { version = ">=0.3.0, <0.4.0", default-features = false, features = ["bundled"] }

[dev-dependencies]
cfg-if = "1"
dotenvy = "0.15"
ipnetwork = ">=0.12.2, <0.22.0"
quickcheck = "1.0.3"
tempfile = "3.10.1"
diesel_test_helper = { path = "../diesel_test_helper" }

[features]
default = ["with-deprecated", "32-column-tables", "mssql", "chrono", "r2d2", "firebird"]
uuid = ["dep:uuid"]
extras = ["chrono", "time", "serde_json", "uuid", "network-address", "numeric", "r2d2"]
unstable = ["diesel_derives/nightly"]
large-tables = ["32-column-tables"]
huge-tables = ["64-column-tables"]
32-column-tables = ["diesel_derives/32-column-tables"]
64-column-tables = ["32-column-tables", "diesel_derives/64-column-tables"]
128-column-tables = ["64-column-tables", "diesel_derives/128-column-tables"]
postgres = ["dep:pq-sys", "postgres_backend"]
sqlite = ["dep:libsqlite3-sys", "dep:sqlite-wasm-rs", "diesel_derives/sqlite", "time?/formatting", "time?/parsing"]
mysql = ["dep:mysqlclient-sys", "dep:url", "dep:percent-encoding", "dep:bitflags", "mysql_backend"]
mssql = ["dep:url", "mssql_backend", "dep:thiserror", "dep:uuid", "dep:bytes", "dep:enumflags2", "dep:byteorder", "dep:tracing", "dep:encoding_rs", "dep:connection-string", "dep:pretty-hex","native-tls","winauth","dep:percent-encoding"]
firebird = ["dep:rsfbclient","chrono"]
native-tls = ["dep:native-tls"]
rustls = ["dep:rustls","dep:rustls-pemfile", "dep:rustls-native-certs"]
bigdecimal = ["dep:bigdecimal"]
mssql_backend = ["diesel_derives/mssql"]
without-deprecated = ["diesel_derives/without-deprecated"]
with-deprecated = ["diesel_derives/with-deprecated"]
network-address = ["dep:ipnetwork", "dep:libc"]
ipnet-address = ["dep:ipnet", "dep:libc"]
numeric = ["dep:num-bigint", "dep:bigdecimal", "dep:num-traits", "dep:num-integer"]
postgres_backend = ["diesel_derives/postgres", "dep:bitflags", "dep:byteorder", "dep:itoa"]
mysql_backend = ["diesel_derives/mysql", "dep:byteorder"]
returning_clauses_for_sqlite_3_35 = []
i-implement-a-third-party-backend-and-opt-into-breaking-changes = []
r2d2 = ["diesel_derives/r2d2", "dep:r2d2"]
chrono = ["diesel_derives/chrono", "dep:chrono", "dep:chrono-tz"]
time = ["diesel_derives/time", "dep:time"]
__with_asan_tests = [
    "libsqlite3-sys?/bundled",
    "libsqlite3-sys?/with-asan",
    "pq-sys?/bundled",
    "pq-src?/with-asan",
    "mysqlclient-sys?/bundled",
    "mysqlclient-src?/with-asan",
]

[package.metadata.docs.rs]
features = ["postgres", "mysql", "sqlite", "mssql", "extras"]
no-default-features = true
rustc-args = ["--cfg", "docsrs"]
rustdoc-args = ["--cfg", "docsrs"]
