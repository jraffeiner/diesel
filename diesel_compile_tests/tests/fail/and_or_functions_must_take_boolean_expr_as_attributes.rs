extern crate diesel;

use diesel::prelude::*;

table! {
    users {
        id -> Integer,
        name -> VarChar,
    }
}

fn main() {
    let conn = &mut PgConnection::establish("â€¦").unwrap();
    users::table
        .filter(users::id.eq(1).and(users::id).or(users::id))
        .select(users::id)
        .execute(conn)
        .unwrap();
}
