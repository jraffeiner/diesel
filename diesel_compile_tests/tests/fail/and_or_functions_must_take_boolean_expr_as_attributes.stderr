error[E0277]: `diesel::sql_types::Integer` is neither `diesel::sql_types::Bool` nor `diesel::sql_types::Nullable<Bool>`
  --> tests/fail/and_or_functions_must_take_boolean_expr_as_attributes.rs:15:33
   |
15 |         .filter(users::id.eq(1).and(users::id).or(users::id))
   |                                 ^^^ the trait `BoolOrNullableBool` is not implemented for `diesel::sql_types::Integer`
   |
   = note: try to provide an expression that produces one of the expected sql types
   = help: the following other types implement trait `BoolOrNullableBool`:
             Bool
             Nullable<Bool>
note: required by a bound in `diesel::BoolExpressionMethods::and`
  --> $DIESEL/src/expression_methods/bool_expression_methods.rs
   |
   |     fn and<T, ST>(self, other: T) -> dsl::And<Self, T, ST>
   |        --- required by a bound in this associated function
...
   |         ST: SqlType + TypedExpressionType + BoolOrNullableBool,
   |                                             ^^^^^^^^^^^^^^^^^^ required by this bound in `BoolExpressionMethods::and`

error[E0277]: `diesel::sql_types::Integer` is neither `diesel::sql_types::Bool` nor `diesel::sql_types::Nullable<Bool>`
  --> tests/fail/and_or_functions_must_take_boolean_expr_as_attributes.rs:15:48
   |
15 |         .filter(users::id.eq(1).and(users::id).or(users::id))
   |                                                ^^ the trait `BoolOrNullableBool` is not implemented for `diesel::sql_types::Integer`
   |
   = note: try to provide an expression that produces one of the expected sql types
   = help: the following other types implement trait `BoolOrNullableBool`:
             Bool
             Nullable<Bool>
note: required by a bound in `diesel::BoolExpressionMethods::or`
  --> $DIESEL/src/expression_methods/bool_expression_methods.rs
   |
   |     fn or<T, ST>(self, other: T) -> dsl::Or<Self, T, ST>
   |        -- required by a bound in this associated function
...
   |         ST: SqlType + TypedExpressionType + BoolOrNullableBool,
   |                                             ^^^^^^^^^^^^^^^^^^ required by this bound in `BoolExpressionMethods::or`
